### PR TITLE
[REFACTOR] Channels: now sync and in context

### PR DESCRIPTION
**Pull request by lucksus**
preparation for #350 (pure reducer refactoring)

Sync channels (https://doc.rust-lang.org/std/sync/mpsc/fn.sync_channel.html) can easily be passed around between threads (which I need for #350).
Since I had to touch every line that mentions channels (or their `Senders`) to change to `SyncSender` and since we have the `context` object holding the state and being passed to everything, I've put those channels into the context.

### CODE REVIEW

File: core/src/context.rs [hunk 1]
```diff
@@ -23,11 +30,32 @@ impl Context {
         logger: Arc<Mutex<Logger>>,
         persister: Arc<Mutex<Persister>>,
     ) -> Context {
+        let (tx_action, _) = sync_channel(100);
+        let (tx_observer, _) = sync_channel(100);
```
Comment (ddd-mtl) [inline core/src/context.rs]: Magic number: Please use a global/struct const instead and possibly add a comment explaining the rationale behind that number.

File: core/src/instance.rs [hunk 2]
```diff
@@ -79,8 +79,8 @@ impl Instance {
 
     /// Returns recievers for actions and observers that get added to this instance
     fn initialize_channels(&mut self) -> (Receiver<ActionWrapper>, Receiver<Observer>) {
-        let (tx_action, rx_action) = channel::<ActionWrapper>();
-        let (tx_observer, rx_observer) = channel::<Observer>();
+        let (tx_action, rx_action) = sync_channel::<ActionWrapper>(100);
+        let (tx_observer, rx_observer) = sync_channel::<Observer>(100);
```
Comment (ddd-mtl) [inline core/src/instance.rs]: magic number

**Comment by ddd-mtl** [comment]
@zippy Yeah I guess `Context` changed from "context of the Holochain instance" to "context of code execution", so the purpose changed but the name still works.

**Comment by thedavidmeister** [comment]
@ddd-mtl @lucksus i'll see if i can clean up the magic number before my plane leaves...

**Comment by lucksus** [comment]
Thanks for fixing the magic numbers, @thedavidmeister!

File: core/src/context.rs [hunk 3]
```diff
@@ -15,19 +20,46 @@ pub struct Context {
...
     pub persister: Arc<Mutex<Persister>>,
     state: Option<Arc<RwLock<State>>>,
+    pub action_channel: SyncSender<ActionWrapper>,
+    pub observer_channel: SyncSender<Observer>,
 }
 
 impl Context {
+    pub fn default_channel_buffer_size() -> usize {
+        100
+    }
```
Comment (sphinxc0re) [inline core/src/context.rs]: Please make this a constant

Comment (thedavidmeister) [reply core/src/context.rs]: @sphinxc0re doesn't the rust compiler optimise this kind of thing away under the hood?

Comment (ddd-mtl) [reply core/src/context.rs]: Yes, I believe so, but I agree with @sphinxc0re that a `const` would be clearer, more scan-able.
I don't see the advantage of this form. I would be happy to learn.

**Pull request merged**